common/http/platform/beast/http.cpp

```diff
@@ -0,0 +1,366 @@
+
+namespace asio = boost::asio;
+namespace http = boost::beast::http;
+
+static http::verb VerbToBeastVerb(Verb verb) {
+	switch (verb) {
+	case GET:
+		return http::verb::get;
+	case POST:
+		return http::verb::post;
+	case PUT:
+		return http::verb::put;
+	case PATCH:
+		return http::verb::patch;
+	case CONNECT:
+		return http::verb::connect;
+	}
+	// Don't use "default" case. This should generate a warning if we ever add any verbs. But
+	// still assert here for safety.
+	assert(false);
+	return http::verb::get;
+}
+
+static Verb BeastVerbToVerb(http::verb verb) {
+	switch (verb) {
+	case http::verb::get:
+		return GET;
+	case http::verb::post:
+		return POST;
+	case http::verb::put:
+		return PUT;
+	case http::verb::patch:
+		return PATCH;
+	case http::verb::connect:
+		return CONNECT;
+	}
+	// Don't use "default" case. This should generate a warning if we ever add any verbs. But
+	// still assert here for safety.
+	assert(false);
+	return GET;
+}
+
+HttpSession::HttpSession(const HttpClient &client, events::EventLoop &event_loop) :
+	logger_("http"),
+	resolver_(GetAsioIoContext(event_loop)),
+	stream_(GetAsioIoContext(event_loop)),
+	body_buffer_(16384) {
+	response_buffer_.reserve(body_buffer_.size());
+}
+
+HttpSession::~HttpSession() {
+	Cancel();
+}
+
+error::Error HttpSession::AsyncCall(
+	HttpRequestPtr req, HttpResponseHandler header_handler, HttpResponseHandler body_handler) {
+	Cancel();
+
+	if (!req->ready_) {
+		return error::MakeError(error::ProgrammingError, "HttpRequest is not ready");
+	}
+
+	if (!header_handler || !body_handler) {
+		return error::MakeError(
+			error::ProgrammingError, "header_handler and body_handler can not be nullptr");
+	}
+
+	if (req->protocol_ != "http") {
+		return error::Error(make_error_condition(errc::protocol_not_supported), req->protocol_);
+	}
+
+	// Use url as context for logging.
+	logger_ = log::Logger("http").WithFields(log::LogField("url", req->orig_address_));
+
+	request_ = req;
+	header_handler_ = header_handler;
+	body_handler_ = body_handler;
+
+	resolver_.async_resolve(
+		request_->host_,
+		to_string(request_->port_),
+		[this](error_code err, const asio::ip::tcp::resolver::results_type &results) {
+			ResolveHandler(err, results);
+		});
+
+	return error::NoError;
+}
+
+void HttpSession::CallErrorHandler(
+	const error_code &err, const HttpRequestPtr &req, HttpResponseHandler handler) {
+	handler(error::Error(
+		err.default_error_condition(), VerbToString(req->method_) + " " + req->orig_address_));
+}
+
+void HttpSession::CallErrorHandler(
+	const error::Error &err, const HttpRequestPtr &req, HttpResponseHandler handler) {
+	handler(error::Error(
+		err.code, err.message + ": " + VerbToString(req->method_) + " " + req->orig_address_));
+}
+
+void HttpSession::ResolveHandler(
+	error_code err, const asio::ip::tcp::resolver::results_type &results) {
+	if (err) {
+		CallErrorHandler(err, request_, header_handler_);
+		return;
+	}
+
+	if (logger_.Level() >= log::LogLevel::Debug) {
```

Comment:
Hmm... Ad-hoc lazy evaluation?

I understand the motivation, but maybe we should think about adding this to the logger lib instead?

Reply:
I have been considering adding a `fmt` lib to the logger, then we could probably support this OOTB.

Reply:
Well, for this particular case the `if` would be needed anyway, because we need to assemble an arbitrary number of results on the same line.

But I do agree that an fmt would be nice in general, and computationally speaking this:

```
log.Debug("%s %s", str1, str2);
```

is much cheaper than this

```
log.Debug(str1 + " " + str2);
```

Because it can check the log level and bail immediately, instead of assembling the strings first and then bailing afterwards.

Reply:
There's `WouldLog(log_level)` in CFEngine. I think it's important to keep the logic in the logging library because it might not be that easy to tell once there are different sinks with different log levels, etc.

Reply:
That's not a bad idea. But I think it's independent of this PR. If we want to change this later it should be pretty easy to grep for `\.Level` and replace this.